src/err.rs

```diff
@@ -171,12 +169,20 @@ impl PyErr {
     ///
     /// The error is cleared from the Python interpreter.
     /// If no error is set, returns a `SystemError`.
-    pub fn fetch(_: Python) -> PyErr {
+    pub fn fetch(py: Python) -> PyErr {
         unsafe {
             let mut ptype: *mut ffi::PyObject = std::ptr::null_mut();
             let mut pvalue: *mut ffi::PyObject = std::ptr::null_mut();
             let mut ptraceback: *mut ffi::PyObject = std::ptr::null_mut();
             ffi::PyErr_Fetch(&mut ptype, &mut pvalue, &mut ptraceback);
+
+            if ptype == PanicException::type_object().as_ptr() {
```

Comment:
I **really** don't understand why we need this.
Why is it problematic to fetch `PanicException`?
cc: @programmerjake

Reply:
The idea is that unless caught a `panic!` should cause a stack unwind all the way until program exit (like it does without this PR). The changes in this PR are intended to make it so that we unwind safely through the FFI layer, because what we do at the moment is technically UB.

This means we need two main things:
   - When Rust code panics, we start unwinding in python by raising `PanicException`.
   - If this `PanicException` is not caught, we should use `std::panic::resume_unwind()` when we get back into Rust code so that we continue the unwind.

I need the original panic payload to call `resume_unwind()`, but at the moment it's difficult for me to store that on `PanicException`. I expect that will create a lot of work which will extend this already large PR, so I opened #853 instead.

Without the original payload, for now I just call `panic!` instead of `std::panic::resume_unwind()`, which is close enough. It has the downside that the panic handler runs again, so at the moment the `thread '<unnamed>' panicked at xyz` message will be printed multiple times as the program unwinds. (Once for the original `panic!` and once each time we continue unwinding when we fetch `PanicException`)

Perhaps I should add a comment to the code documenting this?

Reply:
> because what we do at the moment is technically UB.

It's not true.
As I explained it can cause just a *logic* error.

> If this PanicException is not caught, we should use std::panic::resume_unwind() when we get back into Rust code so that we continue the unwind.

So I'm not still convinced that PyO3 itself should cause panic for that.
Let's consider a scenario where multiple PyO3 crates, say, c1 and c2 work together.
c1 can fetch c2's PanicException. Do we really want to resume panic in such a situation?
How about adding an API (e.g., `resume_panic`) to PanicException?

Reply:
> because what we do at the moment is technically UB.

By this I mean the current pyo3 behaviour of not using `catch_unwind` is UB - Rust panic through the FFI stack frames isn't well defined. This is why we need to add `catch_unwind`.

I agree with you that catching `PanicException` can cause a logic error. This could occur either when Python catches the exception, or if pyO3 does not continue to unwind after fetching a `PanicException`.

This is why I think the default behaviour should be to keep unwinding all the way to program exit - even if c1 fetches c2's `PanicException`.

I'll put some diagrams together this evening to try to explain better.